File: tools/usage-poc/coconut_collector/hashing.py
"""hashing.py — device-local salted project hashing.

projectHash anonymises a project so the same project aggregates stably on
one device while its path/name cannot be recovered off-device.

  projectHash = sha256(salt + ":" + project_slug)[:12]

The salt is a random 32-byte value generated once per device and stored at
~/.coconutlabs/salt (0600). It NEVER leaves the device and is never written
into any Burn Summary. See web/docs/decision/coconutlabs-verification-model.md.
"""

import hashlib
import os
import secrets
from pathlib import Path

SALT_DIR = Path.home() / ".coconutlabs"
SALT_PATH = SALT_DIR / "salt"
_HASH_LEN = 12


def load_or_create_salt(salt_path: Path = SALT_PATH) -> str:
    """Return the device salt (hex), creating it once if absent.

    The salt file is created with 0600 permissions. A pre-existing file is
    left untouched so project hashes stay stable across runs.
    """
    if salt_path.is_file():
        existing = salt_path.read_text(encoding="utf-8").strip()
        if existing:
            # Tighten perms on a pre-existing file: O_CREAT mode never
            # applies to a file that already exists.
            os.chmod(salt_path, 0o600)
            return existing
    salt_path.parent.mkdir(parents=True, exist_ok=True)
    salt = secrets.token_hex(32)
    # Write 0600 atomically: create with restrictive mode, then write.
    fd = os.open(str(salt_path), os.O_WRONLY | os.O_CREAT | os.O_TRUNC, 0o600)
    with os.fdopen(fd, "w", encoding="utf-8") as f:
        f.write(salt)
    # O_CREAT mode is masked by umask; chmod guarantees exactly 0600.
    os.chmod(salt_path, 0o600)
    return salt


def project_hash(project_slug: str, salt: str) -> str:
    """Compute the 12-hex salted hash of a project slug.

    project_slug is hash input only — callers must never emit it raw.
    """
    digest = hashlib.sha256(f"{salt}:{project_slug}".encode("utf-8")).hexdigest()
    return digest[:_HASH_LEN]

File: lib/client/burn/hashing.ts
// hashing.ts — browser-side salted project hashing (Phase 2 FSA flow).
//
// Mirrors web/tools/usage-poc/coconut_collector/hashing.py *semantically*.
// Byte-level parity with Python is intentionally NOT a goal — the browser keeps
// its own salt in IndexedDB so the same machine produces stable hashes across
// browser visits, but its hash space is separate from the Python CLI's by
// default. A user who wants the two paths to collapse into one identity can
// paste their `~/.coconutlabs/salt` value into the optional "Import Python
// salt" input; that's the only way the two paths agree.
//
// Why separate identities by default? `~/.coconutlabs/salt` lives at 0600 on
// the user's home directory — reading it from the browser would require the
// user to grant the page access to their home folder, which the rest of the
// plan explicitly forbids (narrow .claude/projects and .codex/sessions only).
//
// Hash formula (must match hashing.py::project_hash):
//   sha256(TextEncoder().encode(`${salt}:${slug}`))[:12]  // lowercase hex
//
// Salt format (must match hashing.py::load_or_create_salt):
//   secrets.token_hex(32) → 64 lowercase hex characters (32 random bytes).

const DB_NAME = "coconutlabs";
const STORE_NAME = "config";
const SALT_KEY = "salt";
const HASH_LEN = 12;
// Python's secrets.token_hex(32) emits 64 hex chars from 32 random bytes.
const SALT_HEX_LEN = 64;

// Matches the shape of /^[0-9a-f]{64}$/ — what Python's secrets.token_hex(32)
// produces and what an imported Python salt should still look like.
const SALT_RE = /^[0-9a-f]{64}$/;

function openSaltDb(): Promise<IDBDatabase> {
  return new Promise((resolve, reject) => {
    const req = indexedDB.open(DB_NAME, 1);
    req.onupgradeneeded = () => {
      const db = req.result;
      if (!db.objectStoreNames.contains(STORE_NAME)) {
        db.createObjectStore(STORE_NAME);
      }
    };
    req.onsuccess = () => resolve(req.result);
    req.onerror = () => reject(req.error ?? new Error("IndexedDB open failed"));
  });
}

function withStore<T>(
  mode: IDBTransactionMode,
  fn: (store: IDBObjectStore) => IDBRequest<T>,
): Promise<T> {
  return openSaltDb().then(
    (db) =>
      new Promise<T>((resolve, reject) => {
        const tx = db.transaction(STORE_NAME, mode);
        const store = tx.objectStore(STORE_NAME);
        const req = fn(store);
        req.onsuccess = () => resolve(req.result);
        req.onerror = () => reject(req.error ?? new Error("IndexedDB op failed"));
        tx.oncomplete = () => db.close();
        tx.onerror = () => {
          db.close();
          reject(tx.error ?? new Error("IndexedDB tx failed"));
        };
      }),
  );
}

function bytesToHex(bytes: Uint8Array): string {
  let out = "";
  for (let i = 0; i < bytes.length; i++) {
    out += bytes[i].toString(16).padStart(2, "0");
  }
  return out;
}

function generateSalt(): string {
  // 32 random bytes → 64 lowercase hex chars (matches secrets.token_hex(32)).
  const buf = new Uint8Array(32);
  crypto.getRandomValues(buf);
  return bytesToHex(buf);
}

// Read the persisted browser salt, generating one on first use. The salt
// stays IndexedDB-resident — never uploaded, never written into a Burn
// Summary — and is origin-bound by the browser, so coconutlabs.xyz cannot
// see another origin's salt.
export async function loadOrCreateSalt(): Promise<string> {
  const existing = (await withStore("readonly", (s) => s.get(SALT_KEY))) as
    | string
    | undefined;
  if (typeof existing === "string" && SALT_RE.test(existing)) {
    return existing;
  }
  const fresh = generateSalt();
  await withStore("readwrite", (s) => s.put(fresh, SALT_KEY));
  return fresh;
}

// Replace the persisted salt with a user-provided value (the "Import Python
// salt" path). Validates the hex shape before persisting so a typo cannot
// silently corrupt every future hash on this device.
export async function importSalt(rawValue: string): Promise<string> {
  const trimmed = rawValue.trim().toLowerCase();
  if (!SALT_RE.test(trimmed)) {
    throw new Error(
      "Salt must be 64 lowercase hex characters (32 bytes). Open ~/.coconutlabs/salt and copy its contents.",
    );
  }
  await withStore("readwrite", (s) => s.put(trimmed, SALT_KEY));
  return trimmed;
}

// Erase the persisted salt; the next loadOrCreateSalt() call mints a new one.
// Project hashes silently change after this — only call when the user
// explicitly asks to reset their device identity.
export async function clearSalt(): Promise<void> {
  await withStore("readwrite", (s) => s.delete(SALT_KEY));
}

// Byte-equivalent to Python's:
//   hashlib.sha256(f"{salt}:{project_slug}".encode("utf-8")).hexdigest()[:12]
// TextEncoder is UTF-8 native, crypto.subtle.digest returns SHA-256 as raw
// bytes, lowercase hex matches Python's hexdigest, slice(0, 12) matches the
// _HASH_LEN cut.
//
// projectSlug is hash input only — callers must never emit it raw. The slug
// itself is the user's project folder name; the hash is the only form that
// is safe to ship in a Burn Summary.
export async function projectHash(
  projectSlug: string,
  salt: string,
): Promise<string> {
  const input = new TextEncoder().encode(`${salt}:${projectSlug}`);
  const digest = await crypto.subtle.digest("SHA-256", input);
  return bytesToHex(new Uint8Array(digest)).slice(0, HASH_LEN);
}
